### ISSUE TITLE
[Proposal] Generalized Automaton read input API

### ISSUE THREAD

**Issue by khoda81** [caleb531/automata#138]
Mathematically every machine can be defined with a state (which would be the data/memory of the machine) paired with an update function. I propose we break the `read_input_stepwise` method into these methods:
1. `initial_state() -> State`: returns the state that the machine would be at before starting to read the input
2. `delta(state: State, x: Any) -> State`: aka the update function, takes the machine state with a single item from the machine input and returns the new state of the machine
3. `is_final(state: State) -> bool`: returns whether or not the input is accepted if the input were to end here

This can be applied to all machines in this library. A simple function to check if any input is accepted by _any_ automata can be written like this:
```python
def does_accept(automaton: Automaton, input: Iterator[Any]) -> bool:
    state = automaton.initial_state()
    for x in input:
        state = automaton.delta(state, x)

    return automaton.is_final(state)
```

Then the `read_input_stepwise` can be moved to the Automaton class:
```python
def read_input_stepwise(automaton: Automaton, input: Iterator[Any]) -> Generator[State, None, None]:
    state = automaton.initial_state()
    for x in input:
        state = automaton.delta(state, x)
        yield state
```

This will not break any existing API use cases but some new use cases will become available:
 - Searching tree of all the accepted inputs of length less than `n`
 - Provide a default implementation for converting any finite automaton to `DFA`
 - Help with unit testing and visualizations
 - Easier customization of automata, making a custom `FA` that would accept any odd length input can be achieved with a single bool as State and a not operation in the delta function for example.
 - And more...

This API is not yet thoroughly tested and methods are subject to name/signature change. for now, this is an initial idea for the new API.

**Comment by khoda81** [comment]
One missing feature of the current API is to get extra info about a transition. For example, getting edges that were used in a transition can be useful to clean up the step visualization code. Unfortunately, the new API does not fix that either!

It can be achieved by changing the `delta` function to:
- <code>delta(state: State, x: Any) -> **tuple[State, dict]**</code>: takes the machine state with a single item from the machine input and returns the new state of the machine **with some extra info about the transition.**

Then expect `NFA`/`DFA` classes to return a list of edges (tuples of (from, to, label)) contained in the dict. for example:
```python
>>> state, info = fa.delta(state, "0")
>>> info
{
    "edges": [("q0", "q1", "0"), ("q0", "q2", "0")],
    "lambda_closure": {"q1", "q2"}
}
```

**Comment by eliotwrobson** [comment]
@khoda81 I have a few issues with the way this proposed API would interact with the existing code:

1. The main issue is that the transition function `delta(state: State, x: Any) -> State` actually needs to return a set in the case of an NFA. You could always have the DFA return a singleton set, but this could be needlessly confusing. In the followup comment, I'm really not a fan of having a dict with string keys returned, although this is more flexible it makes it hard to annotate things with types and can be done with just a regular tuple.
2. I don't really see the point in having an `initial_state` and `final_state` function when these are already given explicitly as sets as part of the public interface. If this is a push towards implementing automaton more abstractly internally as a collection of functions, I think that's really not a good direction for this library right now (as most of the algorithms rely on explicit definitions of states and transitions).

I think it's a good idea to expand the functionality of `read_input_stepwise` to make visualization easier, but I think the proposed changes here include some things that are out of scope and would be hard to implement.

**Comment by khoda81** [comment]
@eliotwrobson Yes, the delta method's return type depends on the automaton type. Most of the time we don't need to care about the type of state because it is only handled by the three methods and we just pass it around. In case we need to work state, we do checks to make sure we are using the right automaton. Mypy does support type hints like this:
```python
from typing import Any, Generator, Generic, Hashable, Iterable, Mapping, Optional, TypeVar

import abc


State = TypeVar("State")
class Automaton(Generic[State], metaclass=abc.ABCMeta):
    @abc.abstractmethod
    def delta(self, state: State, x: Any) -> State:
        raise NotImplementedError

    @abc.abstractproperty
    def initial_state(self) -> State:
        raise NotImplementedError

    @abc.abstractmethod
    def is_terminal(self, state: State) -> bool:
        raise NotImplemented

    def delta_star(self, input: Iterable[Any], state: Optional[State] = None) -> State:
        input = iter(input)
        state = self.initial_state if state is None else state

        for x in input:
            state = self.delta(state, x)

        return state


class Oddomaton(Automaton[bool]):
    """
    An automaton that accepts any input of odd length.
    """

    initial_state = False

    def delta(self, state: bool, _x: Any) -> bool:
        return not state

    def is_terminal(self, state: bool) -> bool:
        return state


T = TypeVar("T", Hashable)
I = TypeVar("I", Hashable)
class NFA(Generic[T], Automaton[frozenset[T]]):
    def __init__(
        self,
        *,
        states: Iterable[T],
        input_symbols: Iterable[I],
        transitions: Mapping[T, Mapping[I, T]],
        initial_state: Iterable[T],
        final_states: Iterable[T],
    ):
        """Initialize a complete NFA."""
        super().__init__()
        self.states = frozenset(states)
        self.input_symbols = frozenset(input_symbols)
        self.transitions = transitions
        self.final_states = frozenset(final_states)
        self._lambda_closures = self._compute_lambda_closures(states, transitions)

        # change1: we can set initial_state after _lambda_closures
        self.initial_state = self._lambda_closures[initial_state]

    # change2: We should probably inline _get_next_current_states here. And why 
    # is _get_next_current_states private? I can see it being useful at times.
    def delta(self, state: frozenset[T], x: I) -> frozenset[T]:
        return self._get_next_current_states(state, x)

    # change3: This method is also implemented with exceptions instead of 
    # returning a bool.
    def is_terminal(self, state: frozenset[T]) -> bool:
        return state.isdisjoint(self.final_states)


def is_odd(input: Iterable[Any]):
    automaton = Oddomaton()
    state = automaton.delta_star(input)
    return automaton.is_terminal(state)

```
Depending on how specific you say the type of the automaton, Mypy can infer the state automatically.

>  I'm really not a fan of having a dict with string keys returned

Me neither, which is why it is in the next comment. I'm still looking for a nice way of implementing that...

> I don't really see the point in having an `initial_state` and `final_state`

Yes, they exist only for the purpose of abstraction and flexibility. Maybe it's a good idea to make initial_state a property rather than a method. One problem with the current API is that it is mostly built around having string inputs even though it can support iterables of any type with minimal changes.

> proposed changes here include some things that are out of scope and would be hard to implement.

Most of these methods are implemented already, it is just a matter of moving them around.
